libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SFHybridViewController.m

```diff
@@ -446,28 +456,24 @@ - (NSString *)isLoginRedirectUrl:(NSURL *)url
     if (url == nil || url.absoluteString == nil || url.absoluteString.length == 0) {
         return nil;
     }
-    if ([url.scheme.lowercaseString hasPrefix:@"http"]) {
+    if ([url.scheme.lowercaseString hasPrefix:kHTTP]) {
         if (url.query != nil) {
-            NSString *startUrlValue = [url valueForParameterName:@"startURL"];
-            NSString *ecValue = [url valueForParameterName:@"ec"];
-            BOOL foundStartURL = (startUrlValue != nil);
-            BOOL foundValidEcValue = ([ecValue isEqualToString:@"301"] || [ecValue isEqualToString:@"302"]);
-            if (foundValidEcValue) {
-                if (foundStartURL) {
-                    return startUrlValue;
-                } else {
-                    return self.startPage;
-                }
-            } else if ([self isSamlLoginRedirect:url.absoluteString]) {
-                return self.startPage;
+            NSString *startUrlValue = [url valueForParameterName:kStartURLParam];
+            if (startUrlValue == nil || [startUrlValue containsString:kFrontdoor]) {
+                startUrlValue = self.startPage;
+            }
+            if ([self isSessionExpirationRedirect:url.absoluteString] || [self isSamlLoginRedirect:url.absoluteString] || [self isVFPageRedirect:url]) {
+                return startUrlValue;
             }
-        } else if ([self isSamlLoginRedirect:url.absoluteString]) {
-            return self.startPage;
         }
     }
     return nil;
 }
 
+- (BOOL)isSessionExpirationRedirect:(NSString *)url {
+    return (url && [url containsString:kHybridSessionRedirect]);
```

Comment:
That's the part that is new?
Everything else is just refactoring, right?

Reply:
Yes, that's basically it. Check for the redirect, extract `retURL` and load it through `frontdoor`. The rest is just optimizing the code.